File: hw3/code/Assig3Q4.py
import numpy as np
import statsmodels.api as sm
from scipy.stats import t
import matplotlib.pyplot as plt

np.random.seed(1010)

M = 100
lower = np.zeros(M)
upper = np.zeros(M)
x2 = np.random.uniform(0, 20, 50)  

for i in range(M):
    y = 10 + 5 * x2 + np.random.normal(0, 6, 50)
   
    X = sm.add_constant(x2)
   
    model = sm.OLS(y, X).fit()
   
    b2 = model.params[1]
   
    varb2 = model.cov_params()[1, 1]
   
    se2 = np.sqrt(varb2)
   
    t_val = t.ppf(0.975, 48)  
    lower[i] = b2 - t_val * se2
    upper[i] = b2 + t_val * se2

CIs = np.column_stack((lower, upper))

IDg = np.where((lower <= 5) & (upper >= 5))[0]
length_IDg = len(IDg)

IDb = np.where(~((lower <= 5) & (upper >= 5)))[0]
length_IDb = len(IDb)

ratiog = (length_IDg / M) * 100
print("Ratio", ratiog)

plt.figure()
plt.xlim([4, 6])
plt.ylim([0, 100])
plt.xlabel(r'$\beta_2$')
plt.ylabel('Samples')

plt.axvline(x=5, color='black', linestyle='--')

colors = ['gray'] * 100  
colors = np.array(colors)
colors[IDb[IDb < 100]] = 'red' 

for j in range(100):
    plt.plot([CIs[j, 0], CIs[j, 1]], [j, j], color=colors[j], lw=2)

plt.show()

File: hw3/code/Assig3Q4.R
set.seed(1010)
M<-100
lower <- numeric(M) 
upper <- numeric(M)
x2<-runif(50,0,20)

for(i in 1:M) {
  y <- 10+5*x2+rnorm(50, mean = 0, sd = 6) 
  model<-lm(y~x2)
  b2<-coef(model)[[2]]
  varb2 <- vcov(model)[2, 2]
  se2<-sqrt(  varb2)
  lower[i] <- b2 - qt(0.025,48, lower.tail = F) * se2 
  upper[i] <- b2 + qt(0.025,48, lower.tail = F) * se2
}
CIs <- cbind(lower, upper)

IDg <- which((CIs[1:M, 1] <= 5 & 5 <= CIs[1:M, 2]))
length(IDg)
IDb <- which(!(CIs[1:M, 1] <= 5 & 5 <= CIs[1:M, 2]))

ratiog<-(length(IDg)/M)*100
ratiog

plot(0,
     xlim = c(4, 6),
     ylim = c(1, 100),
     ylab = "Samples",
     xlab = expression(beta[2]),
     main = "Confidence Intervals")
abline(v = 5, lty = 2)

colors <- rep(gray(0.6), 100) 
colors[IDb] <- "red"

for(j in 1:100) {
  lines(c(CIs[j, 1], CIs[j, 2]), c(j, j),
        col = colors[j],
        lwd = 2)
}


